## PR #885 Review: Fix `Storage::Fromfile` count handling

*Posted by Claude Code on behalf of @pcchen*

### Overview

Fixes a genuine bug in `Storage::Fromfile`: when `count >= 0`, `Nelem` was never initialized before the comparison `count > Nelem`, making the bounds check undefined behaviour and valid counted reads potentially spurious-throw. The fix computes `total_elements` once before the branch and uses it consistently in both paths.

---

### Correctness

**The bug fix is correct.** Computing `total_elements = Nbytes / Type.typeSize(dtype)` before the branch eliminates the UB. Both paths are now properly defined:

- `count < 0`: `Nelem = total_elements` (unchanged semantics)
- `count >= 0`: compare `requested_count` against `total_elements`, then `Nelem = requested_count`

**Cast safety:** `static_cast<cytnx_uint64>(count)` is safe — the `else` branch is only reached when `count >= 0`, so the signed→unsigned cast is well-defined.

**Format string:** Changing `%d` (which was wrong for a `cytnx_uint64`) to `%llu` with explicit `unsigned long long` casts is correct. The old code was also a latent format-string bug.

---

### Issues

#### 1. `cytnx_error_msg` format: confirm the macro accepts two `%llu` arguments

The old call had one variadic arg; the new call has two:

```cpp
cytnx_error_msg(requested_count > total_elements,
                "[ERROR] count (%llu) exceeds ... (%llu) in file.\n",
                static_cast<unsigned long long>(requested_count),
                static_cast<unsigned long long>(total_elements));
```

`cytnx_error_msg` is variadic, so this should work — but worth confirming that the macro's internal `sprintf`/`snprintf` call passes `__VA_ARGS__` rather than a fixed number of arguments. If it's defined with a single fixed slot, the second argument would be silently dropped.

#### 2. Missing boundary-case tests (minor)

The two new tests cover `count == 2` (prefix) and `count == 5` (overflow). Two additional cases would give complete coverage:

- `count == 0` — empty read; should produce a zero-size Storage
- `count == total_elements` — exact full read with explicit count; should succeed and match the `count < 0` path

Neither is a bug, but `count == 0` in particular exercises a path (`Nelem = 0`) that could interact with downstream allocators.

---

### Test quality

- `RemoveFileOnExit` RAII guard is well-written — cleanup runs even if the test throws. ✅
- `::testing::TempDir()` is the correct way to get a test-scoped temp directory. ✅
- Pre-call `std::remove` before writing is defensive and harmless. ✅
- `ASSERT_EQ(loaded.size(), 2)` before the element checks correctly aborts the test on a size mismatch. ✅

---

### Recommendation

**Approve after confirming** the `cytnx_error_msg` macro passes all variadic arguments to its internal format call (issue 1). The boundary-case tests (issue 2) are optional but encouraged.

--- a/tests/Storage_test.cpp
+++ b/tests/Storage_test.cpp
@@ -275,3 +275,33 @@ TEST_F(StorageTest, FromfileRejectsCountLargerThanFile) {
 
   EXPECT_THROW(Storage::Fromfile(path, Type.Double, 5), std::logic_error);
 }
+
+TEST_F(StorageTest, FromfileCountZeroReturnsEmptyStorage) {
+  const std::string path = ::testing::TempDir() + "cytnx_storage_fromfile_count_zero.bin";
+  std::remove(path.c_str());
+  const RemoveFileOnExit cleanup(path);
+
+  Storage source = Storage::from_vector(std::vector<cytnx_double>{1.0, 2.0, 3.0, 4.0});
+  source.Tofile(path);
+
+  Storage loaded = Storage::Fromfile(path, Type.Double, 0);
+  EXPECT_EQ(loaded.dtype(), Type.Double);
+  EXPECT_EQ(loaded.size(), 0);
+}
+
+TEST_F(StorageTest, FromfileCountEqualsTotalElementsReadsAll) {
+  const std::string path = ::testing::TempDir() + "cytnx_storage_fromfile_count_exact.bin";
+  std::remove(path.c_str());
+  const RemoveFileOnExit cleanup(path);
+
+  Storage source = Storage::from_vector(std::vector<cytnx_double>{1.0, 2.0, 3.0, 4.0});
+  source.Tofile(path);
+
+  // count == total elements: explicit count should match count < 0 behaviour
+  Storage loaded_explicit = Storage::Fromfile(path, Type.Double, 4);
+  Storage loaded_default = Storage::Fromfile(path, Type.Double, -1);
+  ASSERT_EQ(loaded_explicit.size(), 4);
+  ASSERT_EQ(loaded_default.size(), 4);
+  for (cytnx_uint64 i = 0; i < 4; i++)
+    EXPECT_DOUBLE_EQ(loaded_explicit.at<cytnx_double>(i), loaded_default.at<cytnx_double>(i));
+}

